### PR TITLE
[DropZone] Use createRef() API 

### PR DESCRIPTION
**Pull request by whitetig3r**
### WHY are these changes introduced?
These changes capitalize on the enhancements made for refs in >React 16.3.
Resolves https://github.com/Shopify/polaris-react/issues/910

### WHAT is this pull request doing?
The refs to the DOM nodes are now initialized using createRef(). These refs are linked to the respective DOM elements in the render method, through the ref prop. The refs are now accessed using _this.refname.current_. There are no visible UI changes.

### 🎩 checklist

* [x] Tested on [mobile](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md#cross-browser-testing)
* [x] Tested on [multiple browsers](https://help.shopify.com/en/manual/intro-to-shopify/shopify-admin/supported-browsers)
* [x] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)

### CODE REVIEW

File: src/components/DropZone/DropZone.tsx [hunk 1]
```diff
@@ -151,10 +151,10 @@ export class DropZone extends React.Component<CombinedProps, State> {
     return Object.keys(newState).length ? newState : null;
   }
 
-  private node: HTMLElement | null = null;
-  private dropNode: HTMLElement | HTMLDocument | null = null;
+  private node: React.RefObject<HTMLDivElement>;
+  private dropNode: React.RefObject<HTMLDivElement> | Document;
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: Why did you change the type from `HTMLDocument` to `Document`?

File: src/components/DropZone/DropZone.tsx [hunk 2]
```diff
@@ -177,6 +177,9 @@ export class DropZone extends React.Component<CombinedProps, State> {
       errorOverlayText: translate(`Polaris.DropZone.errorOverlayText${suffix}`),
       numFiles: 0,
     };
+    this.node = React.createRef();
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: Rather than creating the refs in the constructor we can do it above

File: src/components/DropZone/DropZone.tsx [hunk 3]
```diff
@@ -151,10 +151,10 @@ export class DropZone extends React.Component<CombinedProps, State> {
     return Object.keys(newState).length ? newState : null;
   }
 
-  private node: HTMLElement | null = null;
-  private dropNode: HTMLElement | HTMLDocument | null = null;
+  private node: React.RefObject<HTMLDivElement>;
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]:
```suggestion
  private node: React.createRef<HTMLDivElement>();
```

File: src/components/DropZone/DropZone.tsx [hunk 4]
```diff
@@ -304,30 +307,58 @@ export class DropZone extends React.Component<CombinedProps, State> {
 
     this.setState({error: this.props.error});
 
-    if (!this.dropNode) {
+    const dropNode: HTMLDivElement | Document | null =
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: Typescript should infer the return type so we can omit it

File: src/components/DropZone/DropZone.tsx [hunk 5]
```diff
@@ -304,30 +307,58 @@ export class DropZone extends React.Component<CombinedProps, State> {
...
+      'current' in this.dropNode ? this.dropNode.current : document;
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: Could we change this to
```tsx
const dropNode = this.dropNode.current || document;
```

File: src/components/DropZone/DropZone.tsx [hunk 6]
```diff
@@ -304,30 +307,58 @@ export class DropZone extends React.Component<CombinedProps, State> {
...
+
+    if (!dropNode) {
       return;
     }
 
-    addEventListener(this.dropNode, 'drop', this.handleDrop);
-    addEventListener(this.dropNode, 'dragover', this.handleDragOver);
-    addEventListener(this.dropNode, 'dragenter', this.handleDragEnter);
-    addEventListener(this.dropNode, 'dragleave', this.handleDragLeave);
+    addEventListener(dropNode as EventTarget, 'drop', this.handleDrop);
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: Any idea why we have to assert the type? Ideally we shouldn't have to force it.

File: src/components/DropZone/DropZone.tsx [hunk 7]
```diff
@@ -304,30 +307,58 @@ export class DropZone extends React.Component<CombinedProps, State> {
...
+
     if (this.props.openFileDialog) {
       this.triggerFileDialog();
     }
   }
 
   componentWillUnmount() {
-    if (!this.dropNode) {
+    const dropNode: HTMLDivElement | Document | null =
+      'current' in this.dropNode ? this.dropNode.current : document;
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: Same as above 😄

File: src/components/DropZone/DropZone.tsx [hunk 8]
```diff
@@ -541,7 +557,12 @@ export class DropZone extends React.Component<CombinedProps, State> {
     }
 
     this.dragTargets = this.dragTargets.filter((el: Node) => {
-      return el !== event.target && this.dropNode && this.dropNode.contains(el);
+      return (
+        el !== event.target &&
+        'current' in this.dropNode &&
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: We can omit this check since we do it below

File: src/components/DropZone/DropZone.tsx (hunk 1 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: Thanks for pointing that out. Will change all usage of `Document` back to `HTMLDocument` 😄

File: src/components/DropZone/DropZone.tsx (hunk 2 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: Will move the refs above!

File: src/components/DropZone/DropZone.tsx (hunk 3 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: Will add these changes! 😄 

eg: `private node = React.createRef<HTMLDivElement>()` like this?

File: src/components/DropZone/DropZone.tsx (hunk 4 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: Yes! 👍  That's right! Thank you! Will omit that check.

File: src/components/DropZone/DropZone.tsx (hunk 5 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: `[ts]
Property 'current' does not exist on type 'RefObject<HTMLDivElement> | Document'.
  Property 'current' does not exist on type 'Document'.`

I think that a type check seems to be imperative in this case, before accessing the current property

File: src/components/DropZone/DropZone.tsx (hunk 6 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: `[ts]
Argument of type 'HTMLDivElement | Document | null' is not assignable to parameter of type 'EventTarget'.
  Type 'null' is not assignable to type 'EventTarget'. `

I guess this has to do with the definitions of the event listeners in `@shopify/javascript-utilities/events` which requires that the first argument be of type _EventTarget_.

Any other workarounds? 😄

File: src/components/DropZone/DropZone.tsx (hunk 7 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: Will remove that check! 😄

File: src/components/DropZone/DropZone.tsx (hunk 8 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: `[ts] Object is possibly 'null'.`

Is the check necessary since,`in` just checks existence while `this.dropNode.current` checks whether it is null before trying to see whether it contains `el` ?

File: src/components/DropZone/DropZone.tsx [hunk 9]
```diff
@@ -152,10 +152,13 @@ export class DropZone extends React.Component<CombinedProps, State> {
     return Object.keys(newState).length ? newState : null;
   }
 
-  private node: HTMLElement | null = null;
-  private dropNode: HTMLElement | HTMLDocument | null = null;
+  private node = React.createRef<HTMLDivElement>();
+  private dropNode = this.props.dropOnPage
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: I don't think we need to use `createRef` here since it's never actually it's own ref. It's either document or `this.node`. This assignment will also only be ran once opposed to every render.

File: src/components/DropZone/DropZone.tsx [hunk 10]
```diff
@@ -327,30 +330,36 @@ export class DropZone extends React.Component<CombinedProps, State> {
 
     this.setState({error: this.props.error});
 
-    if (!this.dropNode) {
+    const dropNode = this.dropNode.current || document;
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: An alternative solution that may fix the problem above is user a `getter`. Rather than a `private property` we can use a dynamic value. Something like the example below may work.

```tsx
private get dropNode() {
  return this.props.dropOnPage && this.node.current ? document : null;
}
```

File: src/components/DropZone/DropZone.tsx [hunk 11]
```diff
@@ -327,30 +330,36 @@ export class DropZone extends React.Component<CombinedProps, State> {
...
-    addEventListener(this.dropNode, 'dragover', this.handleDragOver);
-    addEventListener(this.dropNode, 'dragenter', this.handleDragEnter);
-    addEventListener(this.dropNode, 'dragleave', this.handleDragLeave);
+    addEventListener(dropNode, 'drop', this.handleDrop);
+    addEventListener(dropNode, 'dragover', this.handleDragOver);
+    addEventListener(dropNode, 'dragenter', this.handleDragEnter);
+    addEventListener(dropNode, 'dragleave', this.handleDragLeave);
     addEventListener(window, 'resize', this.adjustSize);
 
+    this.adjustSize();
```
Comment (AndrewMusgrave) [inline src/components/DropZone/DropZone.tsx]: Previously this was being called every render now it's just being called in `componentDidMount`

File: src/components/DropZone/DropZone.tsx (hunk 9 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: Thank you for pointing that out! Yes, it would be quite redundant.

File: src/components/DropZone/DropZone.tsx (hunk 10 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: That's a perfect fix! I've added that change. 😄

File: src/components/DropZone/DropZone.tsx (hunk 11 above)
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: After some study on ref callbacks, I've understood that ref callbacks are called before the component mounts (if the component is rendered) and not on every render. So, with the new refs API, it would be best to call it in ComponentDidMount alone?

**Comment by whitetig3r** [comment]
I've been unable to fresh a review, for some reason. 😅

**Comment by AndrewMusgrave** [comment]
> I've been unable to fresh a review, for some reason.

Weird, it's a new tool so there's bound to be some bugs. I'll refresh myself 😄

File: src/components/DropZone/DropZone.tsx [hunk 12]
```diff
@@ -152,10 +152,13 @@ export class DropZone extends React.Component<CombinedProps, State> {
     return Object.keys(newState).length ? newState : null;
   }
 
-  private node: HTMLElement | null = null;
-  private dropNode: HTMLElement | HTMLDocument | null = null;
+  private node = React.createRef<HTMLDivElement>();
+  private get dropNode() {
```
[the remark this answers is not in the document]
Comment (whitetig3r) [reply src/components/DropZone/DropZone.tsx]: Makes it look neat! 😄 Added that change!

**Comment by whitetig3r** [comment]
> This is looking really good and almost ready to 🚢 Would you mind rebasing with master to resolve the conflicts and squash the commits down 😄

That's awesome! Done rebasing with `master` and squashed down to one commit! 😄

File: UNRELEASED.md [hunk 13]
```diff
@@ -22,4 +22,6 @@ Use [the changelog guidelines](https://git.io/polaris-changelog-guidelines) to f
 
 ### Code quality
 
+- Migrated the refs in `DropZone` to use the new createRef API ([#910](https://github.com/Shopify/polaris-react/issues/910))
```
Comment (AndrewMusgrave) [inline UNRELEASED.md]:
```suggestion
- Migrated the refs in `DropZone` to use the new createRef API ([#1063](https://github.com/Shopify/polaris-react/pull/1063))
```

**Comment by whitetig3r** [comment]
> Sorry this has taken me so long to re-review @whitetig3r 🙏 . I wasn't aware it was ready for review, I'm guessing there are still issues with refreshing review? Looks like there's a conflict again in the changelog, and suggested a small change to match the pull request. Feel free to @ me when you'r ready for this to be merged. Thanks for your contribution 🎉 🎉

That's absolutely fine 😄 . Yes, I guess it's because I'm unable to request a review! Sorry for the inconsistency with the changelog addition. I've resolved that and squashed down for a clean commit history. Thanks a ton for the code-reviews @AndrewMusgrave ! Glad to have been able to contribute! 😄

**Comment by AndrewMusgrave** [comment]
I was going to merge this before more conflicts pop up but `codecov` seems to be stuck, will check back later

**Comment by whitetig3r** [comment]
> I was going to merge this before more conflicts pop up but `codecov` seems to be stuck, will check back later

Oh oh! It's been stuck for quite a while now. Should we wait on it or do a `rebase` and `force push` to make it cycle through all the checks again @AndrewMusgrave ?

**Comment by AndrewMusgrave** [comment]
Probably @whitetig3r , Iv'e checked `codecov` and the task finished but it seems github never received a finished status for the check

**Comment by whitetig3r** [comment]
> Probably @whitetig3r , Iv'e checked `codecov` and the task finished but it seems github never received a finished status for the check

Good to merge or should anything else be done? :smiley:

**Comment by AndrewMusgrave** [comment]
> Good to merge or should anything else be done?

Seemed to be stuck again 😬 so I force merged it. Thanks for your contribution 😄 🎉

**Comment by whitetig3r** [comment]
> > Good to merge or should anything else be done?
> 
> Seemed to be stuck again  so I force merged it. Thanks for your contribution

Happy to have been able to contribute and hoping to contribute more! Thank you for the guidance :smiley:  @AndrewMusgrave
